File: validPalindromeII.py
#easy solution
#https://leetcode.com/problems/valid-palindrome-ii/discuss/107718/Easy-to-Understand-Python-Solution
class Solution:
    def validPalindrome(self, s: str) -> bool:
        l, r = 0, len(s)-1
        def isPalin(a: str) -> bool:
            return a[:] == a[::-1]
        while l < r:
            if s[l] != s[r]:
                #discuss [l:r-1] case and [l+1, r] case
                return isPalin(s[l:r]) or isPalin(s[l+1: r+1])
            l += 1
            r -= 1
        return True

File: validPalindromeII.kt
class Solution {
    fun isPalindrome(s: String): Boolean {
        var l = 0
        var r = s.length -1
        do {
            if (s[l] != s[r]) return false
            l++
            r--
        } while (l < r)
        return true
    }

    fun validPalindrome(s: String): Boolean {
        var l = 0
        var r = s.length -1
        do {
            // check either substring l to r-1 and l+1 to r is palindrome
            if (s[l] != s[r]) return isPalindrome(s.substring(l, r)) || isPalindrome(s.substring(l+1, r+1))
            l++
            r--
        } while (l < r)
        return true
    }
}
